src/grouping_map.rs

```diff
@@ -609,6 +559,241 @@ where
     where
         V: Mul<V, Output = V>,
     {
-        self.reduce(|acc, _, val| acc * val)
+        self.product_in(HashMap::new())
+    }
+}
+
+impl<I, K, V> GroupingMap<I>
+where
+    I: Iterator<Item = (K, V)>,
+{
+    /// Apply [`aggregate`](Self::aggregate) with a provided empty map
+    /// (`BTreeMap` or `HashMap` with any hasher).
+    pub fn aggregate_in<FO, R, M>(self, mut operation: FO, mut map: M) -> M
+    where
+        FO: FnMut(Option<R>, &K, V) -> Option<R>,
+        M: Map<Key = K, Value = R>,
+    {
+        debug_assert!(map.is_empty());
```

Comment:
We need empty maps, but which of the following?
- `map.clear();` More convenient but may be surprising?
- `assert!(map.is_empty());` Maybe it should be checked outside of debug mode.
- `debug_assert!(map.is_empty());`

(Same for `collect_in`).

Reply:
I don't think this assert is really needed. The logic doesn't depend on the fact the map is clear, so I would leave this choice to the user. Allowing non-empty maps would also allow using these methods as `extend`-like variants of the others.